### PR TITLE
Add Send Notification Resolver

### PR DESCRIPTION
**Pull request by markWarr**

### CODE REVIEW

File: HousingRepairsOnlineApi.Tests/HelpersTests/NotificationConfiguration/CommunalNotificationConfigurationProviderTests.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            Assert.True(systemUnderTest.ConfirmationSmsTemplateId == confirmationSmsTemplateId);
+            Assert.True(systemUnderTest.InternalEmailTemplateId == internalEmailTemplateId);
+        }
+
+        //Arrange
+        public static IEnumerable<object[]> InvalidStringArgumentTestData()
+        {
+            yield return new object[] { new ArgumentNullException(), null };
+            yield return new object[] { new ArgumentException(), "" };
+        }
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi.Tests/HelpersTests/NotificationConfiguration/CommunalNotificationConfigurationProviderTests.cs:77]: Should whitespace be included as an invalid argument exception?

File: HousingRepairsOnlineApi.Tests/HelpersTests/NotificationConfiguration/TenantNotificationConfigurationProviderTests.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            Assert.True(systemUnderTest.ConfirmationSmsTemplateId == confirmationSmsTemplateId);
+            Assert.True(systemUnderTest.InternalEmailTemplateId == internalEmailTemplateId);
+        }
+
+        //Arrange
+        public static IEnumerable<object[]> InvalidStringArgumentTestData()
+        {
+            yield return new object[] { new ArgumentNullException(), null };
+            yield return new object[] { new ArgumentException(), "" };
+        }
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi.Tests/HelpersTests/NotificationConfiguration/TenantNotificationConfigurationProviderTests.cs:77]: Should whitespace be included as an invalid argument exception?

File: HousingRepairsOnlineApi.Tests/HelpersTests/NotificationConfigurationResolverTests.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        Dictionary<string, INotificationConfigurationProvider> notificationConfigurationProviders =
+            new Dictionary<string, INotificationConfigurationProvider>();
+        notificationConfigurationProviders.Add("TENANT", tenantNotificationConfigurationProvider.Object);
+
+        systemUnderTest = new NotificationConfigurationResolver(notificationConfigurationProviders);
+    }
+
+    [Fact]
+    public void GivenInvalidRepairType_WhenResolving_ThenExceptionIsReturned()
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi.Tests/HelpersTests/NotificationConfigurationResolverTests.cs:30]: This title indicates this is testing an invalid repair type, but the parameter is `COMMUNAL` which is a valid [repair type](https://github.com/Newarkandsherwood/housing-repairs-online-api/blob/main/HousingRepairsOnlineApi/Helpers/RepairType.cs).

Would it be more appropriate to name this test `Given[Unregistered|Unresolvable]RepairType...`?

I feel that `GivenInvalidRepairType_WhenResolving_ThenExceptionIsReturned` is also a valid test, which would need to pass an invalid argument to the `Resolve` method.

File: HousingRepairsOnlineApi.Tests/HelpersTests/NotificationConfigurationResolverTests.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    public void GivenValidRepairType_WhenResolving_ThenTenantNotificationConfigurationProviderIsReturned()
+    {
+
+        // Act
+        var actual = systemUnderTest.Resolve("TENANT");
+
+        // Assert
+        actual.Should().BeEquivalentTo(tenantNotificationConfigurationProvider.Object);
+    }
+
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi.Tests/HelpersTests/NotificationConfigurationResolverTests.cs:51]: Nitpick: Remove extra blank lines

File: HousingRepairsOnlineApi.Tests/HelpersTests/InternalEmailSenderTests.cs [hunk 5]
```diff
@@ -10,17 +11,35 @@ namespace HousingRepairsOnlineApi.Tests.HelpersTests
...
+
+        private readonly string templateId = "123";
+        private readonly string imgUrl = "image.url";
+
 
         public InternalEmailSenderTests()
         {
             retrieveImageLinkUseCase = new Mock<IRetrieveImageLinkUseCase>();
-            internalEmailSender = new Mock<IInternalEmailSender>();
+            //internalEmailSender = new Mock<IInternalEmailSender>();
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi.Tests/HelpersTests/InternalEmailSenderTests.cs:31]: Uncomment or remove commented code
```suggestion
```

File: HousingRepairsOnlineApi/Helpers/InternalEmailSender.cs [hunk 6]
```diff
@@ -9,24 +9,20 @@ public class InternalEmailSender : IInternalEmailSender
...
-            if (!String.IsNullOrEmpty(repair.Description?.PhotoUrl))
-            {
-                await Task.Run(() =>
-                {
-                    imageLink = retrieveImageLinkUseCase.Execute(repair.Description?.PhotoUrl);
-                });
-            }
-            sendInternalEmailUseCase.Execute(repair.Id, repair.Address.LocationId, repair.Address.Display, repair.SOR, repair.Description.Text, repair.ContactDetails?.Value, imageLink);
+            var sendNotification = notificationConfigurationResolver.Resolve(repair.RepairType);
+            var imageLink = await sendNotification.GetImageLink(retrieveImageLinkUseCase, repair);
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/InternalEmailSender.cs:24]: Was it not possible to utilise `GetImageLink` within `INotificationConfigurationProvider.GetPersonalisationForInternalEmailTemplate` rather than call it here?

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/CommunalNotificationConfigurationProvider.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        Guard.Against.NullOrWhiteSpace(repair.Time.Display, nameof(repair.Time.Display), "The appointment time provided is invalid");
+
+        return new Dictionary<string, dynamic>
+        {
+            {"repair_ref", repair.Id},
+            {"appointment_time", repair.Time.Display}
+        };
+    }
+
+    public async Task<string> GetImageLink(IRetrieveImageLinkUseCase retrieveImageLinkUseCase, Repair repair)
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/NotificationConfiguration/CommunalNotificationConfigurationProvider.cs:66]: This method is repeated twice - once for [tenant](https://github.com/Newarkandsherwood/housing-repairs-online-api/pull/36/files#diff-40e41965529f7261733f92c99f151faafce70af296ec3e434c6d06ca985d3ca5R66) and here for [Communal](https://github.com/Newarkandsherwood/housing-repairs-online-api/pull/36/files#diff-b692aaee1da12c929fe22d63bfb2d0d6075a9d46fe90d5239198b1a17487695cR66) (and again in future for leaseholder too).

Is it possible to reduce this duplication?

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/INotificationConfigurationProvider.cs [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+namespace HousingRepairsOnlineApi.Helpers
+{
+    public interface INotificationConfigurationProvider
+    {
+        public string ConfirmationSmsTemplateId { get; set; }
+        public string ConfirmationEmailTemplateId { get; set; }
+        public string InternalEmailTemplateId { get; set; }
+        public Dictionary<string, dynamic> GetPersonalisationForInternalEmailTemplate(Repair repair);
+        public Dictionary<string, dynamic> GetPersonalisationForEmailTemplate(Repair repair);
+        public Dictionary<string, dynamic> GetPersonalisationForSMSTemplate(Repair repair);
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/NotificationConfiguration/INotificationConfigurationProvider.cs:15]: Nitpick: correct casing
```suggestion
        public Dictionary<string, dynamic> GetPersonalisationForSmsTemplate(Repair repair);
```

File: HousingRepairsOnlineApi/UseCases/ISendAppointmentConfirmationEmailUseCase.cs [hunk 9]
```diff
@@ -1,10 +1,11 @@
...
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
 using HousingRepairsOnlineApi.Domain;
 
 namespace HousingRepairsOnlineApi.UseCases
 {
     public interface ISendAppointmentConfirmationEmailUseCase
     {
-        public void Execute(string email, string bookingRef, string appointmentTime);
+        public void Execute(string email, Dictionary<string, dynamic> personalisation, string templateId);
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/UseCases/ISendAppointmentConfirmationEmailUseCase.cs:9]: Use interface for parameter type 
```suggestion
        public void Execute(string email, IDictionary<string, dynamic> personalisation, string templateId);
```

File: HousingRepairsOnlineApi/Properties/launchSettings.json [hunk 10]
```diff
@@ -36,9 +36,15 @@
         "AZURE_STORAGE_CONNECTION_STRING": "<REPLACE WITH AZURE STORAGE CONNECTION STRING>",
         "STORAGE_CONTAINER_NAME": "housing-repairs-online",
         "GOV_NOTIFY_KEY" : "<REPLACE WITH GOV NOTIFY API URL>",
-        "CONFIRMATION_SMS_NOTIFY_TEMPLATE_ID": "<REPLACE WITH GOV NOTIFY CONFIRMATION SMS TEMPLATE ID>",
-        "CONFIRMATION_EMAIL_NOTIFY_TEMPLATE_ID": "<REPLACE WITH GOV NOTIFY CONFIRMATION EMAIL TEMPLATE ID>",
-        "INTERNAL_EMAIL_NOTIFY_TEMPLATE_ID": "<REPLACE WITH GOV NOTIFY INTERNAL EMAIL TEMPLATE ID>",
+        "TENANT_CONFIRMATION_SMS_NOTIFY_TEMPLATE_ID": "<REPLACE WITH GOV NOTIFY CONFIRMATION SMS TEMPLATE ID>",
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Properties/launchSettings.json:39]: Changing the existing environment variable names and adding new ones will require a Terraform change too, which is not present on this PR.

File: HousingRepairsOnlineApi/Startup.cs [hunk 11]
```diff
@@ -40,6 +42,21 @@ public void ConfigureServices(IServiceCollection services)
                     new EnvironmentVariableRepairTypeSorConfigurationProvider(RepairType.Communal),
                 });
 
+            var sendTenantNotification = new TenantNotificationConfigurationProvider(EnvironmentVariableHelper.GetEnvironmentVariable("TENANT_CONFIRMATION_SMS_NOTIFY_TEMPLATE_ID"),
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Startup.cs:45]: The instantiation of each notification configuration provider within the startup _could_ causing longer startup times (and increased memory usage due to always being present and not managed through Dependancy Injection lifetimes). Was this intentional?

File: HousingRepairsOnlineApi/UseCases/ISendAppointmentConfirmationSmsUseCase.cs [hunk 12]
```diff
@@ -1,10 +1,11 @@
...
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
 using HousingRepairsOnlineApi.Domain;
 
 namespace HousingRepairsOnlineApi.UseCases
 {
     public interface ISendAppointmentConfirmationSmsUseCase
     {
-        public void Execute(string number, string bookingRef, string appointmentTime);
+        public void Execute(string number, Dictionary<string, dynamic> personalisation, string templateId);
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/UseCases/ISendAppointmentConfirmationSmsUseCase.cs:9]: Use interface for parameter type

File: HousingRepairsOnlineApi/UseCases/ISendInternalEmailUseCase.cs [hunk 13]
```diff
@@ -1,10 +1,11 @@
-﻿using System.Threading.Tasks;
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
 
 namespace HousingRepairsOnlineApi.UseCases
 {
     public interface ISendInternalEmailUseCase
     {
-        public void Execute(string repairRef, string uprn, string address, string sor, string repairDescription, string contactNumber, string image);
+        public void Execute(Dictionary<string, dynamic> personalisation, string imagelink, string templateId);
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/UseCases/ISendInternalEmailUseCase.cs:8]: Use interface for parameter type

**Review by Bhavesh-Patel — CHANGES_REQUESTED** [review]

File: HousingRepairsOnlineApi/Helpers/InternalEmailSender.cs (hunk 6 above)
Comment (markWarr) [reply HousingRepairsOnlineApi/Helpers/InternalEmailSender.cs]: I have now moved this into INotificationConfigurationProvider

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/CommunalNotificationConfigurationProvider.cs (hunk 7 above)
Comment (markWarr) [reply HousingRepairsOnlineApi/Helpers/NotificationConfiguration/CommunalNotificationConfigurationProvider.cs]: I've added a base class "BaseNotificationConfigurationProvider" to specify the common behaviour.

File: HousingRepairsOnlineApi/Startup.cs (hunk 11 above)
Comment (markWarr) [reply HousingRepairsOnlineApi/Startup.cs:45]: Yes, I couldn't see another way of doing this, and seemed to fit in with the way we add the SOR engines, which are  created in startup: 

`services.AddSorEngines(
                new[]
                {
                    new EnvironmentVariableRepairTypeSorConfigurationProvider(RepairType.Tenant),
                    new EnvironmentVariableRepairTypeSorConfigurationProvider(RepairType.Communal),
                });`

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+using System.Threading.Tasks;
+using Ardalis.GuardClauses;
+using HousingRepairsOnlineApi.Domain;
+using HousingRepairsOnlineApi.UseCases;
+
+namespace HousingRepairsOnlineApi.Helpers.SendNotifications;
+
+public class BaseNotificationConfigurationProvider
+{
+    public string ConfirmationSmsTemplateId { get; set; }
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs:11]: Each of these properties has a public setter, which doesn't seem to be used.
If they are not required externally of this class make them private (or alternatively use [Init only setters](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/init)).

File: HousingRepairsOnlineApi.Tests/UseCasesTests/SendInternalEmailUseCaseTests.cs [hunk 15]
```diff
@@ -39,9 +39,9 @@ public static IEnumerable<object[]> InvalidImageArgumentTestData()
         public void GivenNoImage_WhenExecute_ThenGovNotifyGateWayIsCalled()
         {
             //Act
-            const string Base64Img = "";
+            //const string Base64Img = "";
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi.Tests/UseCasesTests/SendInternalEmailUseCaseTests.cs:42]: Remove commented code
```suggestion
```

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs [hunk 16]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Ardalis.GuardClauses;
+using HousingRepairsOnlineApi.Domain;
+using HousingRepairsOnlineApi.UseCases;
+
+namespace HousingRepairsOnlineApi.Helpers.SendNotifications;
+
+public class BaseNotificationConfigurationProvider
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs:9]: Nitpick: Given this doesn't provide functionality of its own, consider making it `abstract`.

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/CommunalNotificationConfigurationProvider.cs [hunk 17]
```diff
@@ -51,7 +46,7 @@ public class CommunalNotificationConfigurationProvider : INotificationConfigurat
         };
     }
 
-    public Dictionary<string, dynamic> GetPersonalisationForSMSTemplate(Repair repair)
+    public Dictionary<string, dynamic> GetPersonalisationForSmsTemplate(Repair repair)
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/NotificationConfiguration/CommunalNotificationConfigurationProvider.cs:49]: Use interface for return type
```suggestion
    public IDictionary<string, dynamic> GetPersonalisationForSmsTemplate(Repair repair)
```

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/INotificationConfigurationProvider.cs [hunk 18]
```diff
@@ -10,9 +10,8 @@ public interface INotificationConfigurationProvider
         public string ConfirmationSmsTemplateId { get; set; }
         public string ConfirmationEmailTemplateId { get; set; }
         public string InternalEmailTemplateId { get; set; }
-        public Dictionary<string, dynamic> GetPersonalisationForInternalEmailTemplate(Repair repair);
+        public Task<Dictionary<string, dynamic>> GetPersonalisationForInternalEmailTemplate(Repair repair, IRetrieveImageLinkUseCase retrieveImageLinkUseCase);
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/NotificationConfiguration/INotificationConfigurationProvider.cs:13]: Use interface for return types for this and the following two methods too
```suggestion
        public Task<IDictionary<string, dynamic>> GetPersonalisationForInternalEmailTemplate(Repair repair, IRetrieveImageLinkUseCase retrieveImageLinkUseCase);
```

**Review by Bhavesh-Patel — CHANGES_REQUESTED** [review]
Updating code focused changes

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+
+    protected BaseNotificationConfigurationProvider(string confirmationSmsTemplateId, string confirmationEmailTemplateId,
+        string internalEmailTemplateId)
+    {
+        ConfirmationSmsTemplateId = confirmationSmsTemplateId;
+        ConfirmationEmailTemplateId = confirmationEmailTemplateId;
+        InternalEmailTemplateId = internalEmailTemplateId;
+    }
+
+    protected async Task<string> GetImageLink(Repair repair, IRetrieveImageLinkUseCase retrieveImageLinkUseCase)
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs:23]: This is not related to configuration and I don't think it belongs here - it doesn't use any of the 
It's more suited to an extension method on `Repair`.

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs (hunk 16 above)
Comment (Bhavesh-Patel) [reply HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs:9]: Alternatively, add a [default interface implementation](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods) and remove this base class. As per my other [comment](https://github.com/Newarkandsherwood/housing-repairs-online-api/pull/36/files/7f627cd898da4ee9ad440dc7403c535129269bdd..c74a355a11258c80e69416d74535c8242da831ae#r1026226161), `GetImageLink` doesn't seem related to configuration

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs (hunk 14 above)
Comment (markWarr) [reply HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs:11]: init only setters - what a nice feature. I've implemented it using init only setters. These are needed to be accessed from outside the class.

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs (hunk 16 above)
Comment (markWarr) [reply HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs]: I've made it abstract so now it just sets the initial properties in the constructor. I wasn't sure a default interface implementation could be used here. You can have methods, but can you define the actual properties? I'll have another look once I've made these initial changes.

File: HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs (hunk 14 above)
Comment (markWarr) [reply HousingRepairsOnlineApi/Helpers/NotificationConfiguration/BaseNotificationConfigurationProvider.cs]: So I kept them public but through the getter.

File: HousingRepairsOnlineApi/Helpers/RepairExtensions.cs [hunk 20]
```diff
@@ -0,0 +1,24 @@
+﻿using System.Threading.Tasks;
+using HousingRepairsOnlineApi.Domain;
+using HousingRepairsOnlineApi.UseCases;
+
+
+namespace HousingRepairsOnlineApi.Helpers
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/RepairExtensions.cs:6]: To aid discoverability of extension methods on Repair, the namespace should be the same as the class being extended, i.e. 
```suggestion
namespace HousingRepairsOnlineApi.Domain
```

On that basis, it would also be best to move this file to the `Domain` folder too.

File: HousingRepairsOnlineApi/Helpers/RepairExtensions.cs [hunk 21]
```diff
@@ -0,0 +1,24 @@
+﻿using System.Threading.Tasks;
+using HousingRepairsOnlineApi.Domain;
+using HousingRepairsOnlineApi.UseCases;
+
+
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Helpers/RepairExtensions.cs:5]: Nitpick: Remove extra blank line

**Review by Bhavesh-Patel — APPROVED** [review]
